File: leetcode/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        // O(n) time and O(n) memory
        vector<int> maxLeft(height.size(), 0), maxRight(height.size(), 0);
        int trapped = 0;
        
        for (int i = 1; i < height.size(); i++)
        {
            maxLeft[i] = max(height[i-1], maxLeft[i-1]);
        }
        for (int i = height.size() - 2; i >= 0; i--)
        {
            maxRight[i] = max(height[i+1], maxRight[i+1]);
        }

        for (int i = 0; i < height.size(); i++)
        {
            trapped += max(0, min(maxRight[i], maxLeft[i]) - height[i]);
        }

        return trapped;
    }
};

File: leetcode/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        trapped = 0
        max_left, max_right = [0 for i in range(len(height))], [0 for i in range(len(height))]

        for i in range(1, len(height)):
            max_left[i] = max(max_left[i-1], height[i-1])
        for i in range(len(height) - 2, -1, -1):
            max_right[i] = max(max_right[i+1], height[i+1])
        
        for i in range(len(height)):
            trapped += max(min(max_left[i], max_right[i]) - height[i], 0)
        
        return trapped